### PR TITLE
Allow users to manage set of installed repositories for creating functions.

### LINKED ISSUE #172: Allow users to manage set of installed repositories for creating functions.
Manages template repositories installed on disk.

1. Allow the user to specify the repository location and use the --repository flag. Once added, a template from the repository can be used when creating a new Function.

_Acceptance Criteria_: The Repository Flag: Installing repositories locally is optional. To use a template from a remote repository directly, it is possible to use the --repository flag on create. This leaves the local disk untouched. For example, To create a Function using the Boson Project Hello-World template without installing the template repository locally, use the --repository (-r) flag on create:

- [ ] Add Repository: Add a new repository to the installed set.
- [ ] List Repository: List all available repositories, including the installed default repository.
- [ ] Rename Repository: Rename a previously installed repository from. Only installed repositories can be renamed. 
- [ ] Remove Repository: Remove a repository by name. Removes the repository from local storage entirely. 

Reference: https://github.com/knative-sandbox/kn-plugin-func/blob/main/docs/guides/commands.md#repository

### PR DESCRIPTION
**Pull request by sudhirverma**
Fix: #172

### CODE REVIEW

**Review by lstocchi — CHANGES_REQUESTED** [review]
The wizard doesn't fully convince me. 
1) Any reason why you used a textbox for the repository field? I would have expected a combo so i can select a repo in the list. Then if you want to ask for a URL you should show a URL in repo list, i only see names. I would never remember the url associated to each name honestly.
image

2) At first i was a bit confused because i didn't see any option about repositories (and i was only looking for this feature so i was focussed on it) so i would assume that a normal user who is not specifically looking for it would have missed it easily. Why don't you add a repository field from the beginning if there is more than one repo to choose from?

I would see this like:
```
name
language
template
repository
location
```

the repository is a combo with the `default` repo selected by default. If this is changed, the `template` field become a textfield (in future you could keep it as a combo and update the items based on the repo)

**Comment by sudhirverma** [comment]
> the repository is a combo with the `default` repo selected by default. If this is changed, the `template` field become a textfield (in future you could keep it as a combo and update the items based on the repo)

`repository` is a `git repository` URL  not a name and I will keep it optional while creating the function.

```
apple@Apples-MacBook-Pro ~ % func repository add boson https://github.com/boson-project/templates
apple@Apples-MacBook-Pro ~ % func repository list 
default
boson
apple@Apples-MacBook-Pro ~ % func create -l go \
--template hello-world \
--repository boson func-name     
Error: path does not exist: boson 
```

and jeff has created https://github.com/knative-sandbox/kn-plugin-func/issues/874 Issue when it is fixed I will show it as a list in the template and  language

**Comment by mohitsuman** [comment]
@sudhirverma There is a merged PR from func cli around this https://github.com/knative-sandbox/kn-plugin-func/pull/1114

Can yo u try and see if it helps solving the use case.

**Review by lstocchi — CHANGES_REQUESTED** [review]
Looks good. Everything works fine.

My concerns are about two things.
The first is the location of the actions to work with repos. Right now they are in the root's context menu. Also in IJ we had this behavior and, with Jeff, we decided to move it on the toolbar.
I wonder if we could do something similar as these are actions which will be rarely used and we can save space for more important actions in future. 
The idea could be adding an icon in the toolbar for everything which is about repository and when selected, a quickpick shows users all the different actions they can perform (add, remove, rename, list)
Furthermore i would allow to invoke them from the command palette. 
But this is just my idea, up to you if you think it's better the current way as Ij and Vscode are quite different.

The second is the `enable git repo` field in the new func wizard. Its behaviour is strange..

image

When you select the language, the template combo gets filled with all the alternatives you can choose from (taken from the default repo and other repos added previously), so why a user would want to use another registry without adding it as a new repo?  
Then if i click on the field it shows me the already added repos as options, but the template combo already provide me the chance to select all templates from that repo. So it would not make sense to create a new func this way, no?

File: src/cli/func-api.ts [hunk 1]
```diff
@@ -91,6 +107,26 @@ export class FuncAPI {
     return funcCliCommand(runCommand);
   }
 
+  static addRepository(name: string, URL: string, namespace: string): CliCommand {
+    const addRepositoryCommand = ['repository', 'add', name, URL, '-n', namespace];
+    return funcCliCommand(addRepositoryCommand);
+  }
+
+  static listRepository(): CliCommand {
+    const listRepositoryCommand = ['repository', 'list', '-v'];
```
Comment (lstocchi) [inline src/cli/func-api.ts:116]: Shouldn't it have the `-n` flag?

**Review by lstocchi — APPROVED** [review]
LGTM, just a missing flag if i'm not wrong. For the rest, it works fine, GJ!! :+1:

File: src/cli/func-api.ts (hunk 1 above)
Comment (sudhirverma) [reply src/cli/func-api.ts]: done thanks :)

**Pull request merged**